dbms/src/Interpreters/ExternalLoader.cpp

```diff
@@ -441,10 +372,149 @@ void ExternalLoader::reload(const std::string & name)
+            else
+                failed_loadable_objects.emplace(name, std::move(info));
+
+            std::rethrow_exception(exception_ptr);
+        }
+        else if (object_ptr->supportUpdates())
+        {
+            const auto & lifetime = object_ptr->getLifetime();
+            if (lifetime.min_sec != 0 && lifetime.max_sec != 0)
+            {
+                std::uniform_int_distribution<UInt64> distribution(lifetime.min_sec, lifetime.max_sec);
+
+                update_times[name] = std::chrono::system_clock::now() +
+                                     std::chrono::seconds{distribution(rnd_engine)};
+            }
+        }
+
+        std::lock_guard lock{map_mutex};
+
+        /// add new loadable object or update an existing version
+        if (object_it == std::end(loadable_objects))
+            loadable_objects.emplace(name, LoadableInfo{std::move(object_ptr), config_path, {}});
+        else
+        {
+            if (object_it->second.loadable)
+                object_it->second.loadable.reset();
+            object_it->second.loadable = std::move(object_ptr);
+
+            /// erase stored exception on success
+            object_it->second.exception = std::exception_ptr{};
+            failed_loadable_objects.erase(name);
+        }
+    }
+    catch (...)
+    {
+        if (!name.empty())
+        {
+            /// If the loadable object could not load data or even failed to initialize from the config.
+            /// - all the same we insert information into the `loadable_objects`, with the zero pointer `loadable`.
+
+            std::lock_guard lock{map_mutex};
+
+            const auto exception_ptr = std::current_exception();
+            const auto loadable_it = loadable_objects.find(name);
+            if (loadable_it == std::end(loadable_objects))
+                loadable_objects.emplace(name, LoadableInfo{nullptr, config_path, exception_ptr});
+            else
+                loadable_it->second.exception = exception_ptr;
+        }
+
+        tryLogCurrentException(log, "Cannot create " + object_name + " '"
+                                    + name + "' from config path " + config_path);
+
+        /// propagate exception
+        if (throw_on_error)
+            throw;
+    }
+}
+
+
 ExternalLoader::LoadablePtr ExternalLoader::getLoadableImpl(const std::string & name, bool throw_on_error) const
 {
-    std::lock_guard lock{map_mutex};
+    /// We try to finish the reloading of the object `name` here, before searching it in the map `loadable_objects` later in this function.
+    /// If some other thread is already doing this reload work we don't want to wait until it finishes, because it's faster to just use
+    /// the current version of this loadable object. That's why we use try_lock() instead of lock() here.
+    std::unique_lock all_lock{all_mutex, std::defer_lock};
+    if (all_lock.try_lock())
+        finishReload(name, throw_on_error);
```

Comment:
Looks like if no other thread which is reloading dictionaries, we will wait. (It's not true).
May be just check that dictionaries have loaded once?
Or first check that dictionary is not in loadable_objects, and then try lock.

Reply:
`try_lock()` never waits, it never blocks the current thread. If `all_mutex` is locked by another thread `try_lock()` just returns false without blocking.

Reply:
Yes. It just looks like in case no other thread locks `all_mutex`, and we successfully locked, all dictionaries will be reloading.

Reply:
Ok, successfully lock is only possible in updating thread. Probably, your solution is better. (system reload dictionary will also resolve dependences fine)